### PR TITLE
Left turns with intermediate goals + several bug fixes

### PR DESCRIPTION
**Pull request by evamo0508**
### New Feature
- able to turn left when no lines are detected (turn left a bit at a time until lines are visible again)
- created a new world for left turn: `state_machine_left.world`
- added a new image topic `/lane/debug` for visualization of the lane detection as well as the ROI. Please update your local `.rivz` if necessary, helps to understand what's going on in the system a lot.
- Insert another argument when launching the state machine, indicating which turn to test. Namely,

```
roslaunch autovalet_system state_machine.launch simulation:=true turn:=left
```
where turn could be either `left` or `right`, and it's defaulted to `right`.
- the front camera is now tilted 10 degrees downward and parameters are tuned accordingly (feature by @SubramanianKrish )

### Bug Fixes
- remove the yellow lines at the edges of the parking lot to avoid lane misdetection.
- ensure the goal generated by goal gen is staying inside the costmap
- ensure the first parking goal won't be published unless it's in the costmap already. Before that, keep moving forward with goals generated by goal gen.
- disable the line tracker since it is actually causing problems (lines lingering when no lines are in FOV)

### What's left
- goal should not only stay inside the costmap but at a fixed portion inside it.
- lines of parking spots being detected as center lines.
- inconsistent pose of the second parking goal

### CODE REVIEW

**Comment by kob51** [comment]
oh so the new left turning logic requires the camera being tilted down?

File: autovalet_description/urdf/autovalet.urdf.xacro [hunk 1]
```diff
@@ -7,7 +7,7 @@
 	<xacro:include filename="$(find velodyne_description)/urdf/VLP-16.urdf.xacro"/>
 
 	<sensor_d435 parent="front_bumper_link" name="frontCamera" >
-	    <origin xyz="-0.305 0.02 0.91" rpy="0 0 0" />
+	    <origin xyz="-0.305 0.02 0.91" rpy="0 0.174 0" />
```
Comment (kob51) [inline autovalet_description/urdf/autovalet.urdf.xacro:10]: this is only 10 deg fyi, not 45 like you said in the description

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py [hunk 2]
```diff
@@ -44,30 +46,65 @@ def generate_goal_from_egoline(self, ego_line, egoline_frame):
...
         _, _, Vt         = np.linalg.svd(ego_line - egoline_midpoint)
         line_direction   = Vt[0]    # The principal direction of the distribution. Line direction here
-        
+
         # Correct sign to ensure that the direction is always away from the robot
         # The fix is to ensure the z-direction is always positive which means direction away from the robot
         line_direction = np.sign(line_direction[-1])*line_direction
 
+        # generate target pos
+        dist_to_costmap  = egoline_midpoint[2] - 7.0
```
Comment (kob51) [inline autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py]: can you maybe read this value (7) from the parameter server? should be `/move_base/global_costmap/height` or width im not sure which. this way allows us to change yaml file without changing code

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py [hunk 3]
```diff
@@ -44,30 +46,65 @@ def generate_goal_from_egoline(self, ego_line, egoline_frame):
...
+    def generate_goal_for_left_turn(self, husky_frame):
+        '''
+        Input:
+        husky_frame   : [string] specifies the name of the frame in which Husky is in
+
+        Output:
+        goal_pose       : [PoseStamped] contains the goal pose in the goal_frame
+        '''
+        target_pos = np.array([3.0, 1.5, 0.0])
+        target_quaternion = quaternion_from_euler(0, 0, np.pi/8)
```
Comment (kob51) [inline autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py:92]: just clarifying, your target is 3m in front and 1.5 m to the left?

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py [hunk 4]
```diff
@@ -69,8 +70,12 @@ def pub_two_goals(self):
         self.goal1 = self.generate_parking_goal(self.tag_tf, pos1, rot1)
         while not parking:
             rospy.sleep(0.5)
+            # keep moving w/ goal gen until goal1 is in costmap
+            if self.dist_to_goal(self.goal1) > 4.0:
```
Comment (kob51) [inline autovalet_husky/src/autovalet_husky/find_parking_spot.py]: should this number (4) also be read off parameter server?

File: autovalet_system/src/autovalet_system.py [hunk 5]
```diff
@@ -133,21 +134,28 @@ def registered_image_callback(self, color_msg, depth_msg):
...
     def sendGoal(self):
         # generate goal from the egoline
         # if((rospy.get_time() - self.previous_time) > 2 and self.ego_line is not None):
         if self.ego_line is not None:
             self.current_goal = self.goalGenerator.generate_goal_from_egoline(self.ego_line, self.depth_frame_id)
             self.goal_pub.publish(self.current_goal)
             self.previous_time = rospy.get_time()
+            self.is_turning_left = False
+            return True
+        elif not self.is_turning_left and self.prev_state != State.START:
```
Comment (kob51) [inline autovalet_system/src/autovalet_system.py]: trying to trace this -- what happens if we're ALREADY turning left and ego line is none

File: autovalet_description/urdf/autovalet.urdf.xacro (hunk 1 above)
Comment (evamo0508) [reply autovalet_description/urdf/autovalet.urdf.xacro:10]: changed my description.

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py (hunk 3 above)
Comment (evamo0508) [reply autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py:92]: correct

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py [hunk 6]
```diff
@@ -44,30 +46,65 @@ def generate_goal_from_egoline(self, ego_line, egoline_frame):
         '''
 
         # ref: https://stackoverflow.com/questions/2298390/fitting-a-line-in-3d/2333251#2333251
+        ego_line         = ego_line[ego_line[:, 2].argsort()]
```
Comment (SubramanianKrish) [inline autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py]: why are we sorting here?

File: autovalet_goal_generation/src/autovalet_goal_generation/utils.py [hunk 7]
```diff
@@ -88,8 +88,8 @@ def make_pose_stamped(frame_id, position, quaternion):
     pose_msg.pose.position.z = position[2]
 
     # Populate the orientation as quaternion
-    pose_msg.pose.orientation.y = quaternion[1]
     pose_msg.pose.orientation.x = quaternion[0]
+    pose_msg.pose.orientation.y = quaternion[1]
```
Comment (SubramanianKrish) [inline autovalet_goal_generation/src/autovalet_goal_generation/utils.py:92]: lol. OCD

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py (hunk 4 above)
Comment (SubramanianKrish) [reply autovalet_husky/src/autovalet_husky/find_parking_spot.py]: +1 I think these better be read off the server

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py [hunk 8]
```diff
@@ -69,8 +70,12 @@ def pub_two_goals(self):
...
+                continue
             # keep publishing the first goal until the robot is close enough
-            if self.dist_to_first_goal(self.goal1) > 0.5:
+            self.first_goal_close = True
+            if self.dist_to_goal(self.goal1) > 0.5:
```
Comment (SubramanianKrish) [inline autovalet_husky/src/autovalet_husky/find_parking_spot.py]: Technically these should be class members. self.first_goal_tol = 0.5 etc. Upto you if you wanna fix them!

File: autovalet_system/scripts/slam.sh [hunk 9]
```diff
@@ -1,8 +1,14 @@
 #!/bin/bash
 sim=$1
-if [ "$sim" = "true" ] 
+turn=$2
+if [ "$sim" = "true" ]
 then
-    roslaunch autovalet_gazebo setup_slam_test.launch world_file:=state_machine.world x:=0 y:=6 1>> $HOME/slam_log.txt 2>&1
+    if [ "$turn" = "right" ]
```
Comment (SubramanianKrish) [inline autovalet_system/scripts/slam.sh:6]: Nicely done!

File: autovalet_system/src/autovalet_system.py (hunk 5 above)
Comment (SubramanianKrish) [reply autovalet_system/src/autovalet_system.py]: I think then this fn returns false in that case?

**Review by SubramanianKrish — COMMENTED** [review]
Looks mostly ok. Feel free to read comments and change anything else. Will approve it when the conversation gets resolved!

File: autovalet_system/src/autovalet_system.py (hunk 5 above)
Comment (kob51) [reply autovalet_system/src/autovalet_system.py]: @SubramanianKrish prove this to me? wonder if this is the cause of the bug we were discussing

**Comment by evamo0508** [comment]
> oh so the new left turning logic requires the camera being tilted down?

This is for the overall better line detection in simulation I believe.

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py (hunk 6 above)
Comment (evamo0508) [reply autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py]: forgot to remove. I was initially trying to take the one-third point along the ego line.

File: autovalet_system/src/autovalet_system.py (hunk 5 above)
Comment (evamo0508) [reply autovalet_system/src/autovalet_system.py]: I changed the logic - only when there are consecutive 5 frames w/o lines detected will a left goal be published.
That being said, if after a left goal is published, there are another 5 frames with no lines, another left goal will be published again.
lmk if the new code makes sense.

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py (hunk 4 above)
Comment (evamo0508) [reply autovalet_husky/src/autovalet_husky/find_parking_spot.py]: done

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py (hunk 2 above)
Comment (evamo0508) [reply autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py]: done

File: autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py (hunk 6 above)
Comment (evamo0508) [reply autovalet_goal_generation/src/autovalet_goal_generation/goal_generator.py]: removed.

**Comment by evamo0508** [comment]
Pushed a new commit. At least on my machine, the left-turn goal doesn't get published when it shouldn't be with the new logic. I am not sure if this is machine-dependent tho....

File: autovalet_husky/src/autovalet_husky/find_parking_spot.py (hunk 8 above)
Comment (evamo0508) [reply autovalet_husky/src/autovalet_husky/find_parking_spot.py]: fixed

**Pull request merged**